File: 计算机工程实践/Transbot SE/pi/src/catkin_ws/src/learning_tf/script/turtle_tf_broadcaster.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import roslib
roslib.load_manifest('learning_tf')
import rospy

import tf
import turtlesim.msg

def handle_turtle_pose(msg, turtlename):
    br = tf.TransformBroadcaster()#定义一个tf广播
    #广播world与输入命名的turtle之间的tf变换
    br.sendTransform((msg.x, msg.y, 0),
                     tf.transformations.quaternion_from_euler(0, 0, msg.theta),
                     rospy.Time.now(),
                     turtlename,
                     "world")

if __name__ == '__main__':
        
    rospy.init_node('turtle1_turtle2_tf_broadcaster')#初始化ros节点
    
    turtlename = rospy.get_param('~turtle') #从参数服务器中获取turtle的名字
    #订阅/pose话题数据，也就是turtle的位姿信息
    rospy.Subscriber('/%s/pose' % turtlename,
                     turtlesim.msg.Pose,
                     handle_turtle_pose,
                     turtlename)
    rospy.spin()

File: 计算机工程实践/Transbot SE/pi/src/catkin_ws/src/learning_tf/src/turtle_tf_broadcaster.cpp
#include <ros/ros.h>
#include <tf/transform_broadcaster.h>
#include <turtlesim/Pose.h>

std::string turtle_name;

void poseCallback(const turtlesim::PoseConstPtr& msg)
{
    
    static tf::TransformBroadcaster br;// 创建tf的广播器

    // 初始化tf数据
    tf::Transform transform;
    transform.setOrigin( tf::Vector3(msg->x, msg->y, 0.0) );//设置xyz坐标
    tf::Quaternion q;
    q.setRPY(0, 0, msg->theta);//设置欧拉角：以x轴，y轴，z轴旋转
    transform.setRotation(q);

    br.sendTransform(tf::StampedTransform(transform, ros::Time::now(), "world", turtle_name));// 广播world与turtle坐标系之间的tf数据
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "turtle_world_tf_broadcaster");// 初始化ROS节点

    if (argc != 2)
    {
        ROS_ERROR("Missing a parameter as the name of the turtle!"); 
        return -1;
    }
    
    turtle_name = argv[1];// 输入参数作为海龟的名字

    // 订阅海龟的位姿话题/pose
    ros::NodeHandle node;
    ros::Subscriber sub = node.subscribe(turtle_name+"/pose", 10, &poseCallback);

        // 循环等待回调函数
    ros::spin();

    return 0;
};
